### DISCUSSION TITLE
Issue with adding columns to action button when combining multiple forms

### DISCUSSION THREAD

**Discussion by unknown1337** [Budibase/budibase#4461] [Help] [answered]
**Goal:** a single webpage submission shall add rows to three data tables
**Current screen design**: 
1. form A
     2. form B
         ...  3. form C
         ..........  Button X

**Current action button X configuration** 
1. validate form A
2. validate form B
3. validate form C
4. Save Row {Data Source= Form A, table = ...}
5. Save Row {Data Source= Form B, table = ...}
6. Save Row {Data Source= Form C, table = ...}

**Issue**: issuing "add Column" to action button Save Row 4 will add that same column to each of the save row actions (4,5,6) while I would expect this extra column to be only added to Save Row 4.

Could this be a bug or should I setup something differently?

Thanks!

**Comment by aptkingston** [comment]


**Reply by unknown1337** [reply]
great, thanks, I'm curious to your result :)

**Reply by aptkingston** [reply]
Hey @unknown1337. This issue turned out to only be visual - the actual data sent to the server was always correct. However, when looking at the button action UI it sometimes looked like data had been overridden by other "save row" actions, even though it had not. So this was just a UI bug, but it looked confusing!

I've opened a PR to fix this: https://github.com/Budibase/budibase/pull/4464

**Comment by shogunpurple** [comment]
Hi @unknown1337 - I see there you have nested forms. We actually prevent this in the UI, but I've noticed that you can bypass it by dragging the forms inside each other. This is a bug - we should not allow for this kind of setup because the nested form contexts can overwrite each other and cause issues. I'll convert this into an issue for us to add more protection in the UI to prevent direct form nesting. 

The other issue that you mention is a problem too - I was just able to recreate it by creating several of the same form actions on different tables - they do overwrite each other every time you update the latest one. 

What you will have to do for now is use 3 separate forms with their own validation and rules. Sorry for the inconvenience!

**Reply by unknown1337** [reply]
thanks @shogunpurple @aptkingston  for your quick feedback! 

@shogunpurple., I can go for three separate forms (all in the same level/depth). Though I need a single submit button (as the three form add row actions have a relation). I think I'm not able to use a **single** submit button to validate and add (new rows) to all three data tables if the forms are not nested (as the button has to be a 'indented child' of a parent form in order to be able to submit a "add row action" @ click). 

Am I correct? if so, then I think this is something that is undesired. If I'm incorrect, I'm looking forward to how you think I should have solved it (now that this action will be prevented in the UI). thanks!

**Comment by unknown1337** [comment, marked as answer]
Friendly gentle reminder as I'm still struggling with this :)

Is (how) it possible to by a single submit button action save rows to multiple tables? Thanks!

**Reply by mjashanks** [reply]
Hi @unknown1337 

In the video below, I show how you can save 2 rows, to 2 separate tables, using one form. Also my app export is below, in case you want to take a closer look.

The key points
1. You use one form
2. You might need to give each field on the form **custom** field names, this is only required if your two tables share a column of the same name. In my case - my `Contact` and `Customer` tables both have a `name` field. So I give my inputs the field names `customer_name` and `contact_name`
3. In the `Save Row` actions for both tables, I manually map each input to its corresponding column.

https://user-images.githubusercontent.com/3524181/154690715-ff122481-b963-4577-a10e-feaab2f7cb6e.mp4

[One Form many tables-export-1645190579968.txt](https://github.com/Budibase/budibase/files/8097482/One.Form.many.tables-export-1645190579968.txt)

**Reply by unknown1337** [reply]
@mjashanks , thanks. That solves my issue. Though I now realize that in your "New Contact" I would like to link (relation e.g. 1:1) to the newCustomerId. Thus: after the saveRow for (new) Customer has finished I'd like to retrieve the new customer ID and also store it in the new Contact row. Am I correct that it is not possible to retreive the new customer ID before submitting the seconds new contact, row? thanks again!

**Reply by mjashanks** [reply]
Yes, so in the 2nd `Save Row` action, one of the binding options should be `{{ Action 1.Result }}` - this returns the row that was saved in the first step.

So, you can use `{{ Action 1.Result._id }}` - and map that to the relationship column.

**Reply by unknown1337** [reply]
thanks again @mjashanks ! How could I have found this myself? (or is that impossible?:P).
